ocaml-lsp-server/src/hover_req.ml

```diff
@@ -12,33 +12,63 @@ let environment_mode =
   | Some true -> Extended_variable
   | Some false | None -> Default
 
-let format_contents ~syntax ~markdown ~typ ~doc =
+let print_dividers sections =
+  match sections with
+  | [] -> ""
+  | _ ->
+    String.concat ~sep:"\n\n --- \n\n" sections
+
+let format_as_code_block strings =
+  match strings with
+  | [] -> ""
+  | _ -> sprintf "```\n %s \n```" (String.concat ~sep:" " strings)
+
+
+let format_contents ~syntax ~markdown ~typ ~doc ~(syntax_documentation:[`Found of Query_protocol.syntax_doc_result
+| `No_documentation ]) ~syntax_doc_setting =
   (* TODO for vscode, we should just use the language id. But that will not work
      for all editors *)
-  `MarkupContent
-    (if markdown then
-       let value =
-         let markdown_name = Document.Syntax.markdown_name syntax in
-         match doc with
-         | None -> sprintf "```%s\n%s\n```" markdown_name typ
-         | Some s ->
-           let doc =
-             match Doc_to_md.translate s with
-             | Raw d -> sprintf "(** %s *)" d
-             | Markdown d -> d
-           in
-           sprintf "```%s\n%s\n```\n---\n%s" markdown_name typ doc
-       in
-       { MarkupContent.value; kind = MarkupKind.Markdown }
-     else
-       let value =
-         match doc with
-         | None -> sprintf "%s" typ
-         | Some d -> sprintf "%s\n%s" typ d
-       in
-       { MarkupContent.value; kind = MarkupKind.PlainText })
+  let syn_doc = match 
+    (Env_vars._ENABLE_SYNTAX_DOCUMENTATION (), syntax_doc_setting) with 
+    | _, true | Some true, _  -> 
+      begin match 
+        syntax_documentation with 
+        | `Found syntax_documentation -> 
+            print_dividers [syntax_documentation.name^": "^syntax_documentation.description; 
+                            "Go to [Documentation]("^syntax_documentation.documentation^")"]
+        | `No_documentation -> ""
+      end
+    | Some false, false | None, false -> ""
+  in
+     `MarkupContent 
+    (if markdown then 
+     let value =
+       let markdown_name = Document.Syntax.markdown_name syntax in
+       let type_info = format_as_code_block [markdown_name; typ] in
+       match doc with
+       | None -> sprintf "%s" (print_dividers [type_info; syn_doc])
```

Comment:
Same here, you can keep `doc` as an option and use `let doc = Option.map doc ...`.

Then at the end, to print all non-None info with the required dividers you can simply filter the list:

```
print_dividers (List.filter_opt [type_info; syntax_doc; doc])
```